### ISSUE TITLE
Sync HydroShare irods userspace on JupyterHub notebook server

### ISSUE THREAD

**Issue by Castronova** [hydroshare/hydroshare#1368]
Currently the JupyterHub notebook server uses the HydroShare REST API for both downloading data for analysis and uploading it back into HydroShare.  I would like to update this approach to leverage the HydroShare irods userspace and `icommands` for: (1) faster upload and download speeds, (2) syncing with irods userspace, (3) enabling all JupyterHub data to be accessed from the **Create Resource** page, and (4) simplifying resource saving/creation from JupyterHub.  Below is a  list of the functional requirements for achieving this:

2. Prepare the jupyterhub docker image
    - [x] Install fuse
    - [x] Install icommands
    - [x] Modify docker host configuration to allow fuse mounting

2. When the **Open With > JupyterHub** button is selected, the current resource's content should be copied into the irods userspace and a JupyterHub notebook should launched.
    - [ ] Need a HydroShare REST API function to move resource content into irods userspace.
    - [ ] Modify the JupyterHub REST handler to invoke this HS REST function.

3. Automatically mount HydroShare userspace when the docker container is instantiated.
    - [ ] Mount irods userspace using `irodsFs`.
        - We need to be sure that **every** hydroshare user has an irods userspace setup.
        - Can we prepare this userspace for everyone using their HydroShare credentials?

4. Update JupyterHub functions to operate on fuse mounted data instead of data downloaded via REST.
    - [ ] Update `hs_utils.py`
    - [ ] Update `jupyterhub_rest_handlers`

5. Save data back into HydroShare
    - [ ] Need a REST API function for moving irods userzone content into new or existing HydroShare resources
    - [ ] Add a button to the JupyterHub notebook that performs saving to HydroShare (`custom.js` and `custom.css`)

**Comment by dtarb** [comment]
A couple of refinements I would like to suggest.

At step 2 when the resource content is copied to the iRODS userspace only the contents folder in the heirarchy <GUID>/data/contents needs to be copied.  I suggest they be copied into a folder <jupyter_sessionDateTime>.  There needs to be a way for a user to see and manage files in this userspace.  I suggest a special userspace resource that lets the user see and then clean up/delete files once the work is done.

At step 3 fuse mounting to be replaced by NFS mounting when available.

At step 5 the save button should only be exposed if the user has write (edit or owner) privileges on the resource.  There should be a save as button to create a new HydroShare resource for cases where the user may not have write permissions, or wants to create a separate resource with the new work that creates a new HydroShare resource with the files in the contents folder.  The save as button should trigger copying of metadata (authors, contributors, keywords etc) from the old resource to the new resource and also leave a "created from" reference to point back to the old resource.

The save or save as buttons also need to set the bag false flag and trigger any needed actions on files of known types that have file level metadata in the resource.

I would be interested in the thoughts of @hyi and @rayi113 from an iRODS perspective @alvacouch from a system architecture perspective and @horsburgh from a resource data model perspective.

**Comment by horsburgh** [comment]
@dtarb - why would we only copy the stuff in the content folder?  What if you need to do something with the manifest of files, or you want to traverse the resource map - and is it ever a good idea to separate the data content from the metadata? The resource is a package - why not just copy the whole package?  You couldn't do a "Save as" that copies metadata for a resource if you only copy the resource's "contents" folder to iRODs.  The metadata file is not stored in the "contents" folder it's in the "data" folder.

**Comment by dtarb** [comment]
The reason I suggested only expose the contents folder is that I do not think users working in a Jupyter notebook environment should change the bagit or metadata files.  If they appear in the directory that has been replicated in the users working environment they are exposed to changing or deleting, thereby corrupting the resource.  I was writing from the assumption that the content files are the users business, but that the other files are HydroShare's business.  

Since you have the original resource GUID and the save as function would need to be an API call, there is access to the metadata for a save as.

**Comment by horsburgh** [comment]
@dtarb - if the actions in the Jupyter Notebook modify the resource files in any way, then the bagit manifest, the resource map document, and the science metadata may also have to be updated because the manifest, dates, and checksums will be changed.  If this is all handled by the "save" call to the API, then that works too.

**Comment by hyi** [comment]
@horsburgh @dtarb Yes, the "save" call to the API will update manifest, dates, and checksums. However, the resource bag zipfile will not be immediately updated but only with bag_modified flag set to True, so the bag zip file is only updated when the bag is being downloaded.

**Comment by aphelionz** [comment]
@Castronova Should this stay here or get moved to the jupyterhub repo?

**Comment by Castronova** [comment]
@aphelionz this can be moved.
